File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/projection/ProjectionUtils.java
package hu.kdea.szavazas.ballotprocessor.projection;

import boofcv.struct.image.GrayU8;
import hu.kdea.szavazas.ballotprocessor.GridConstants;
import hu.kdea.szavazas.ballotprocessor.common.Rect;
import kotlin.Pair;

public final class ProjectionUtils {
    private ProjectionUtils() {
    }

    public static Pair<Integer, Integer> computeRowBoundaries(GrayU8 binary, Rect roi) {
        float[] projection = verticalProjection(binary, roi);
        int top = PeakFinder.findMaxPeak(projection, 0, projection.length / 2);
        int bottom = PeakFinder.findMaxPeak(projection, projection.length / 2, projection.length - 1);
        int cropTop = Math.max(0, top + GridConstants.BOUNDARY_MARGIN);
        int cropBottom = Math.min(projection.length - 1, bottom - GridConstants.BOUNDARY_MARGIN);
        return new Pair<>(cropTop, cropBottom);
    }

    public static float[] columnProjection(GrayU8 binary, Rect roi, int cropTop, int cropBottom) {
        float[] projection = new float[roi.getWidth()];
        for (int x = 0; x < roi.getWidth(); x++) {
            double sum = 0.0;
            for (int y = cropTop; y <= cropBottom; y++) {
                sum += binary.get(roi.getX() + x, roi.getY() + y);
            }
            projection[x] = (float) sum;
        }
        return projection;
    }

    public static float[] rowProjection(GrayU8 binary, Rect roi, int cropTop, int croppedHeight) {
        float[] projection = new float[croppedHeight];
        for (int y = 0; y < croppedHeight; y++) {
            int sum = 0;
            for (int x = 0; x < roi.getWidth(); x++) {
                sum += binary.get(roi.getX() + x, roi.getY() + cropTop + y);
            }
            projection[y] = (float) sum;
        }
        return projection;
    }

    private static float[] verticalProjection(GrayU8 binary, Rect roi) {
        float[] projection = new float[roi.getHeight()];
        for (int y = 0; y < roi.getHeight(); y++) {
            int sum = 0;
            for (int x = 0; x < roi.getWidth(); x++) {
                sum += binary.get(roi.getX() + x, roi.getY() + y);
            }
            projection[y] = (float) sum;
        }
        return projection;
    }
}

File: test-utils/src/main/java/hu/kdea/szavazas/ballotprocessor/projection/ProjectionUtils.kt
package hu.kdea.szavazas.ballotprocessor

import boofcv.struct.image.GrayU8

object ProjectionUtils {

    fun computeRowBoundaries(binary: GrayU8, roi: Rect): Pair<Int, Int> {
        val proj = verticalProjection(binary, roi)
        val top = PeakFinder.findMaxPeak(proj, 0, proj.size / 2)
        val bottom = PeakFinder.findMaxPeak(proj, proj.size / 2, proj.size - 1)
        val cropTop = maxOf(0, top + GridConstants.BOUNDARY_MARGIN)
        val cropBottom = minOf(proj.size - 1, bottom - GridConstants.BOUNDARY_MARGIN)
        return cropTop to cropBottom
    }

    fun columnProjection(binary: GrayU8, roi: Rect, cropTop: Int, cropBottom: Int): FloatArray {
        val proj = FloatArray(roi.width)
        for (x in 0 until roi.width) {
            var sum = 0.0
            for (y in cropTop..cropBottom) {
                sum += binary.get(roi.x + x, roi.y + y)
            }
            proj[x] = sum.toFloat()
        }
        return proj
    }

    fun rowProjection(binary: GrayU8, roi: Rect, cropTop: Int, croppedHeight: Int): FloatArray {
        val proj = FloatArray(croppedHeight)
        for (y in 0 until croppedHeight) {
            var sum = 0
            for (x in 0 until roi.width) {
                sum += binary.get(roi.x + x, roi.y + cropTop + y)
            }
            proj[y] = sum.toFloat()
        }
        return proj
    }

    private fun verticalProjection(binary: GrayU8, roi: Rect): FloatArray {
        val proj = FloatArray(roi.height)
        for (y in 0 until roi.height) {
            var sum = 0
            for (x in 0 until roi.width) {
                sum += binary.get(roi.x + x, roi.y + y)
            }
            proj[y] = sum.toFloat()
        }
        return proj
    }
}